yajl: 2.1.0 -> unstable-2022-04-20

Kiskae opened the pull request:
## Description of changes

1. Changes source to https://github.com/containers/yajl since the original source hasn't seen any updates since 2015
2. Add a patch that strips out the shared library cmake target inside `pkgsStatic`
Fixes #257241
3. Add `testers.testMetaPkgConfig`

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Kiskae commented:
Result of `nixpkgs-review pr 257470` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>diffoscope</li>
    <li>diffoscope.dist</li>
    <li>diffoscope.man</li>
  </ul>
</details>
<details>
  <summary>163 packages built:</summary>
  <ul>
    <li>appvm</li>
    <li>autotiling</li>
    <li>autotiling.dist</li>
    <li>buildah</li>
    <li>buildah.man</li>
    <li>collectd</li>
    <li>cri-o</li>
    <li>cri-o.man</li>
    <li>crun</li>
    <li>docker-machine-kvm</li>
    <li>docker-machine-kvm2</li>
    <li>flashfocus</li>
    <li>flashfocus.dist</li>
    <li>gnome.gnome-boxes</li>
    <li>grive2</li>
    <li>guestfs-tools</li>
    <li>i3</li>
    <li>i3-balance-workspace</li>
    <li>i3-balance-workspace.dist</li>
    <li>i3-layout-manager</li>
    <li>i3-resurrect</li>
    <li>i3-resurrect.dist</li>
    <li>i3-rounded</li>
    <li>i3-rounded.debug</li>
    <li>i3-swallow</li>
    <li>i3-swallow.dist</li>
    <li>i3-wk-switch</li>
    <li>i3.debug</li>
    <li>i3a</li>
    <li>i3a.dist</li>
    <li>i3altlayout</li>
    <li>i3altlayout.dist</li>
    <li>i3status</li>
    <li>io</li>
    <li>jftui</li>
    <li>kcat</li>
    <li>kitti3</li>
    <li>kitti3.dist</li>
    <li>kodi (kodiPackages.kodi)</li>
    <li>kodi-gbm</li>
    <li>kodi-wayland</li>
    <li>kodiPackages.a4ksubtitles</li>
    <li>kodiPackages.archive_tool</li>
    <li>kodiPackages.iagl</li>
    <li>kodiPackages.inputstream-adaptive</li>
    <li>kodiPackages.inputstream-ffmpegdirect</li>
    <li>kodiPackages.inputstream-rtmp</li>
    <li>kodiPackages.invidious</li>
    <li>kodiPackages.joystick</li>
    <li>kodiPackages.kodi-platform</li>
    <li>kodiPackages.libretro</li>
    <li>kodiPackages.libretro-2048</li>
    <li>kodiPackages.libretro-fuse</li>
    <li>kodiPackages.libretro-genplus</li>
    <li>kodiPackages.libretro-mgba</li>
    <li>kodiPackages.libretro-nestopia</li>
    <li>kodiPackages.libretro-snes9x</li>
    <li>kodiPackages.netflix</li>
    <li>kodiPackages.pvr-hdhomerun</li>
    <li>kodiPackages.pvr-hts</li>
    <li>kodiPackages.pvr-iptvsimple</li>
    <li>kodiPackages.steam-controller</li>
    <li>kodiPackages.vfs-libarchive</li>
    <li>kodiPackages.vfs-rar</li>
    <li>kodiPackages.vfs-sftp</li>
    <li>kodiPackages.visualization-fishbmc</li>
    <li>kodiPackages.visualization-goom</li>
    <li>kodiPackages.visualization-matrix</li>
    <li>kodiPackages.visualization-pictureit</li>
    <li>kodiPackages.visualization-projectm</li>
    <li>kodiPackages.visualization-shadertoy</li>
    <li>kodiPackages.visualization-spectrum</li>
    <li>kodiPackages.visualization-starburst</li>
    <li>kodiPackages.visualization-waveform</li>
    <li>krunvm</li>
    <li>libguestfs</li>
    <li>libguestfs-with-appliance</li>
    <li>libmodsecurity</li>
    <li>libmodsecurity.dev</li>
    <li>libvirt</li>
    <li>libvirt-glib</li>
    <li>libvirt-glib.dev</li>
    <li>libvirt-glib.devdoc</li>
    <li>lua51Packages.lua-yajl</li>
    <li>lua52Packages.lua-yajl</li>
    <li>lua53Packages.lua-yajl</li>
    <li>lua54Packages.lua-yajl</li>
    <li>luajitPackages.lua-yajl</li>
    <li>matrix-synapse</li>
    <li>matrix-synapse-plugins.matrix-synapse-ldap3</li>
    <li>matrix-synapse-plugins.matrix-synapse-ldap3.dist</li>
    <li>matrix-synapse-plugins.matrix-synapse-mjolnir-antispam</li>
    <li>matrix-synapse-plugins.matrix-synapse-mjolnir-antispam.dist</li>
    <li>matrix-synapse-plugins.matrix-synapse-s3-storage-provider</li>
    <li>matrix-synapse-plugins.matrix-synapse-s3-storage-provider.dist</li>
    <li>matrix-synapse-plugins.matrix-synapse-shared-secret-auth</li>
    <li>matrix-synapse-plugins.matrix-synapse-shared-secret-auth.dist</li>
    <li>matrix-synapse-unwrapped</li>
    <li>matrix-synapse-unwrapped.dist</li>
    <li>mgmt</li>
    <li>minikube</li>
    <li>minishift</li>
    <li>mpd</li>
    <li>mpd-small</li>
    <li>mpd-small.doc</li>
    <li>mpd.doc</li>
    <li>mpd.man</li>
    <li>mudlet</li>
    <li>multipass</li>
    <li>nwg-displays</li>
    <li>nwg-displays.dist</li>
    <li>nwg-panel</li>
    <li>nwg-panel.dist</li>
    <li>nwg-wrapper</li>
    <li>nwg-wrapper.dist</li>
    <li>ocamlPackages.ocaml_libvirt</li>
    <li>perl536Packages.SysVirt</li>
    <li>perl536Packages.SysVirt.devdoc</li>
    <li>perl538Packages.SysVirt</li>
    <li>perl538Packages.SysVirt.devdoc</li>
    <li>podman</li>
    <li>podman.man</li>
    <li>polybarFull</li>
    <li>powerline (python310Packages.powerline)</li>
    <li>powerline.dist (python310Packages.powerline.dist)</li>
    <li>python310Packages.guestfs</li>
    <li>python310Packages.guestfs.dist</li>
    <li>python310Packages.i3ipc</li>
    <li>python310Packages.i3ipc.dist</li>
    <li>python310Packages.ijson</li>
    <li>python310Packages.ijson.dist</li>
    <li>python310Packages.libvirt</li>
    <li>python310Packages.libvirt.dist</li>
    <li>python310Packages.py3status</li>
    <li>python310Packages.py3status.dist</li>
    <li>python310Packages.pysnow</li>
    <li>python310Packages.pysnow.dist</li>
    <li>python311Packages.guestfs</li>
    <li>python311Packages.guestfs.dist</li>
    <li>python311Packages.i3ipc</li>
    <li>python311Packages.i3ipc.dist</li>
    <li>python311Packages.ijson</li>
    <li>python311Packages.ijson.dist</li>
    <li>python311Packages.libvirt</li>
    <li>python311Packages.libvirt.dist</li>
    <li>python311Packages.pysnow</li>
    <li>python311Packages.pysnow.dist</li>
    <li>raiseorlaunch</li>
    <li>raiseorlaunch.dist</li>
    <li>rubyPackages.ruby-libvirt</li>
    <li>rubyPackages_3_0.ruby-libvirt</li>
    <li>rubyPackages_3_2.ruby-libvirt</li>
    <li>rubyPackages_3_3.ruby-libvirt</li>
    <li>sway-contrib.inactive-windows-transparency</li>
    <li>tengine</li>
    <li>vagrant</li>
    <li>virt-manager</li>
    <li>virt-manager-qt</li>
    <li>virt-manager.dist</li>
    <li>virt-top</li>
    <li>virt-viewer</li>
    <li>xkb-switch-i3</li>
    <li>yajl</li>
  </ul>
</details>

Kiskae commented:
Diffoscope build failure appears to be unrelated: #256896

alyssais commented on pkgs/development/libraries/yajl/default.nix:
```diff
@@ -1,18 +1,26 @@
   };
 
+  patches = lib.optionals stdenv.hostPlatform.isStatic [
+    ./disable_shared.patch
```
Can we not fix upstream to not generate shared libraries when BUILD_SHARED_LIBS is off?

Kiskae replied:
Considering the amount of open PRs including those about CVEs on the 
https://github.com/lloyd/yajl repository I doubt that would go anywhere, but the containers fork might work.

Although just turning off the shared libraries might not be enough, would a `-lyail` accept `libyail_s.a` as a valid link target?

Kiskae replied:
Turns out it was easier than I though: https://github.com/containers/yajl/pull/1

Will amend PR to fetch the patch from upstream.

Kiskae commented:
Swapped patch with upstream PR. This PR also removes `libyajl_s.a` from the default lib output. 

Also enabled the `test` target. There is also a `test-api` target but that relies on a specific folder structure and does not work.

Kiskae commented:
Result of `nixpkgs-review pr 257470` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>diffoscope</li>
    <li>diffoscope.dist</li>
    <li>diffoscope.man</li>
  </ul>
</details>
<details>
  <summary>163 packages built:</summary>
  <ul>
    <li>appvm</li>
    <li>autotiling</li>
    <li>autotiling.dist</li>
    <li>buildah</li>
    <li>buildah.man</li>
    <li>collectd</li>
    <li>cri-o</li>
    <li>cri-o.man</li>
    <li>crun</li>
    <li>docker-machine-kvm</li>
    <li>docker-machine-kvm2</li>
    <li>flashfocus</li>
    <li>flashfocus.dist</li>
    <li>gnome.gnome-boxes</li>
    <li>grive2</li>
    <li>guestfs-tools</li>
    <li>i3</li>
    <li>i3-balance-workspace</li>
    <li>i3-balance-workspace.dist</li>
    <li>i3-layout-manager</li>
    <li>i3-resurrect</li>
    <li>i3-resurrect.dist</li>
    <li>i3-rounded</li>
    <li>i3-rounded.debug</li>
    <li>i3-swallow</li>
    <li>i3-swallow.dist</li>
    <li>i3-wk-switch</li>
    <li>i3.debug</li>
    <li>i3a</li>
    <li>i3a.dist</li>
    <li>i3altlayout</li>
    <li>i3altlayout.dist</li>
    <li>i3status</li>
    <li>io</li>
    <li>jftui</li>
    <li>kcat</li>
    <li>kitti3</li>
    <li>kitti3.dist</li>
    <li>kodi (kodiPackages.kodi)</li>
    <li>kodi-gbm</li>
    <li>kodi-wayland</li>
    <li>kodiPackages.a4ksubtitles</li>
    <li>kodiPackages.archive_tool</li>
    <li>kodiPackages.iagl</li>
    <li>kodiPackages.inputstream-adaptive</li>
    <li>kodiPackages.inputstream-ffmpegdirect</li>
    <li>kodiPackages.inputstream-rtmp</li>
    <li>kodiPackages.invidious</li>
    <li>kodiPackages.joystick</li>
    <li>kodiPackages.kodi-platform</li>
    <li>kodiPackages.libretro</li>
    <li>kodiPackages.libretro-2048</li>
    <li>kodiPackages.libretro-fuse</li>
    <li>kodiPackages.libretro-genplus</li>
    <li>kodiPackages.libretro-mgba</li>
    <li>kodiPackages.libretro-nestopia</li>
    <li>kodiPackages.libretro-snes9x</li>
    <li>kodiPackages.netflix</li>
    <li>kodiPackages.pvr-hdhomerun</li>
    <li>kodiPackages.pvr-hts</li>
    <li>kodiPackages.pvr-iptvsimple</li>
    <li>kodiPackages.steam-controller</li>
    <li>kodiPackages.vfs-libarchive</li>
    <li>kodiPackages.vfs-rar</li>
    <li>kodiPackages.vfs-sftp</li>
    <li>kodiPackages.visualization-fishbmc</li>
    <li>kodiPackages.visualization-goom</li>
    <li>kodiPackages.visualization-matrix</li>
    <li>kodiPackages.visualization-pictureit</li>
    <li>kodiPackages.visualization-projectm</li>
    <li>kodiPackages.visualization-shadertoy</li>
    <li>kodiPackages.visualization-spectrum</li>
    <li>kodiPackages.visualization-starburst</li>
    <li>kodiPackages.visualization-waveform</li>
    <li>krunvm</li>
    <li>libguestfs</li>
    <li>libguestfs-with-appliance</li>
    <li>libmodsecurity</li>
    <li>libmodsecurity.dev</li>
    <li>libvirt</li>
    <li>libvirt-glib</li>
    <li>libvirt-glib.dev</li>
    <li>libvirt-glib.devdoc</li>
    <li>lua51Packages.lua-yajl</li>
    <li>lua52Packages.lua-yajl</li>
    <li>lua53Packages.lua-yajl</li>
    <li>lua54Packages.lua-yajl</li>
    <li>luajitPackages.lua-yajl</li>
    <li>matrix-synapse</li>
    <li>matrix-synapse-plugins.matrix-synapse-ldap3</li>
    <li>matrix-synapse-plugins.matrix-synapse-ldap3.dist</li>
    <li>matrix-synapse-plugins.matrix-synapse-mjolnir-antispam</li>
    <li>matrix-synapse-plugins.matrix-synapse-mjolnir-antispam.dist</li>
    <li>matrix-synapse-plugins.matrix-synapse-s3-storage-provider</li>
    <li>matrix-synapse-plugins.matrix-synapse-s3-storage-provider.dist</li>
    <li>matrix-synapse-plugins.matrix-synapse-shared-secret-auth</li>
    <li>matrix-synapse-plugins.matrix-synapse-shared-secret-auth.dist</li>
    <li>matrix-synapse-unwrapped</li>
    <li>matrix-synapse-unwrapped.dist</li>
    <li>mgmt</li>
    <li>minikube</li>
    <li>minishift</li>
    <li>mpd</li>
    <li>mpd-small</li>
    <li>mpd-small.doc</li>
    <li>mpd.doc</li>
    <li>mpd.man</li>
    <li>mudlet</li>
    <li>multipass</li>
    <li>nwg-displays</li>
    <li>nwg-displays.dist</li>
    <li>nwg-panel</li>
    <li>nwg-panel.dist</li>
    <li>nwg-wrapper</li>
    <li>nwg-wrapper.dist</li>
    <li>ocamlPackages.ocaml_libvirt</li>
    <li>perl536Packages.SysVirt</li>
    <li>perl536Packages.SysVirt.devdoc</li>
    <li>perl538Packages.SysVirt</li>
    <li>perl538Packages.SysVirt.devdoc</li>
    <li>podman</li>
    <li>podman.man</li>
    <li>polybarFull</li>
    <li>powerline (python310Packages.powerline)</li>
    <li>powerline.dist (python310Packages.powerline.dist)</li>
    <li>python310Packages.guestfs</li>
    <li>python310Packages.guestfs.dist</li>
    <li>python310Packages.i3ipc</li>
    <li>python310Packages.i3ipc.dist</li>
    <li>python310Packages.ijson</li>
    <li>python310Packages.ijson.dist</li>
    <li>python310Packages.libvirt</li>
    <li>python310Packages.libvirt.dist</li>
    <li>python310Packages.py3status</li>
    <li>python310Packages.py3status.dist</li>
    <li>python310Packages.pysnow</li>
    <li>python310Packages.pysnow.dist</li>
    <li>python311Packages.guestfs</li>
    <li>python311Packages.guestfs.dist</li>
    <li>python311Packages.i3ipc</li>
    <li>python311Packages.i3ipc.dist</li>
    <li>python311Packages.ijson</li>
    <li>python311Packages.ijson.dist</li>
    <li>python311Packages.libvirt</li>
    <li>python311Packages.libvirt.dist</li>
    <li>python311Packages.pysnow</li>
    <li>python311Packages.pysnow.dist</li>
    <li>raiseorlaunch</li>
    <li>raiseorlaunch.dist</li>
    <li>rubyPackages.ruby-libvirt</li>
    <li>rubyPackages_3_0.ruby-libvirt</li>
    <li>rubyPackages_3_2.ruby-libvirt</li>
    <li>rubyPackages_3_3.ruby-libvirt</li>
    <li>sway-contrib.inactive-windows-transparency</li>
    <li>tengine</li>
    <li>vagrant</li>
    <li>virt-manager</li>
    <li>virt-manager-qt</li>
    <li>virt-manager.dist</li>
    <li>virt-top</li>
    <li>virt-viewer</li>
    <li>xkb-switch-i3</li>
    <li>yajl</li>
  </ul>
</details>

alyssais commented:
Until the patch is accepted or the PR is closed, it's best to still have the patch file in Nixpkgs directly, because if you at some point force push to the PR branch, GitHub will GC the old commits, and then we won't be able to fetch the patch any more.

Kiskae commented:
@alyssais inlined the patch

alyssais commented:
Could you please split the changes into multiple commits?  The commit messages should explain the reason behind changes, like what you've put in the PR body.

Kiskae committed:
yajl: 2.1.0 -> unstable-2022-04-20

Upstream hasn't been updated since 2015.
The 'containers' organization appears to actively use their own fork.

Kiskae committed:
yajl: fix static compilation

Cherry-pick PR that separates compilation of shared and static libraries.

Kiskae committed:
yajl: enable included and pkg-config tests

Kiskae commented:
Done, I considered moving the package to by-name, but I'm not the maintainer so I decided to leave it until the official migration.

alyssais commented on pkgs/development/libraries/yajl/default.nix:
```diff
@@ -1,14 +1,14 @@
 { lib, stdenv, fetchFromGitHub, cmake }:
 
-stdenv.mkDerivation rec {
+stdenv.mkDerivation (finalAttrs: {
```
This isn't used in this commit.

Kiskae replied:
Was working on this in parallel with the `composefs` PR and forgot that this one isn't pinned to a tagged commit. Will fix in a minute.

alyssais reviewed (approved):
Now just needs a rebase to pick up #258548.

Kiskae commented:
I did see a lot of evaluation errors lately, rebased.

saschagrunert commented on pkgs/development/libraries/yajl/cmake-shared-static-fix.patch:
```diff
@@ -0,0 +1,89 @@
+ # set the output path for libraries
+ SET(LIBRARY_OUTPUT_PATH ${libDir})
+ 
+-ADD_LIBRARY(yajl_s STATIC ${SRCS} ${HDRS} ${PUB_HDRS})
```
Can't we build static yajl anymore then?

alyssais replied:
Not as it stands, except through pkgsStatic.  If there's a way in CMake to do shared if possible, and static always, that'd be fine.

alyssais replied:
I see you're working on a project in the same org as the yajl fork we're now using, so the best thing to do would be to work upstream to have the "containers" yajl fork do whatever the best thing to do is, so we don't need to patch here.  (Make sure it still works with `pkgsStatic`.)

Kiskae replied:
Even outside of `pkgsStatic`, adding `-DBUILD_SHARED_LIBS=OFF` as a cmake flag should make it generate static libs. It'd just be named `libyajl.a` instead of `libyajl_s.a`

saschagrunert replied:
Thank you for the clarification folks!
